Source/CarthageKit/Xcode.swift

```diff
@@ -1225,26 +1285,50 @@ public func buildInDirectory(directoryURL: NSURL, withOptions options: BuildOpti
 					} else {
 						filteredSDKs = sdks.filter { options.platforms.contains($0.platform) }
 					}
-
 					return sdkFilter(sdks: filteredSDKs, scheme: scheme, configuration: configuration, project: project)
 				}
-
-				let buildProgress = buildScheme(scheme, withConfiguration: options.configuration, inProject: project, workingDirectoryURL: directoryURL, derivedDataPath: options.derivedDataPath, toolchain: options.toolchain, sdkFilter: wrappedSDKFilter)
-					// Discard any existing Success values, since we want to
+				
+				return buildScheme(scheme, withOptions: options, inProject: project, workingDirectoryURL: directoryURL, dependency: dependency, sdkFilter: wrappedSDKFilter)
+					.on(started: {
+						observer.sendNext(.Success(initialValue))
+					})
+			}
+			.startWithSignal({ (signal, signalDisposable) in
+				disposable += signalDisposable
+				
+				let ignoredValue = (ProjectLocator.Workspace(NSURL(string: "")!), "")
+				
+				let eventSignal: Signal<TaskEvent<(ProjectLocator, String)>, CarthageError> = signal
+					// Discard any Success values, since we want to
 					// use our initial value instead of waiting for
 					// completion.
-					.map { taskEvent in
-						return taskEvent.map { _ in initialValue }
+					.map { (taskEvent: TaskEvent<NSURL>) -> TaskEvent<(ProjectLocator, String)> in
+						return taskEvent.map { _ in ignoredValue }
 					}
-					.filter { taskEvent in taskEvent.value == nil }
-
-				return BuildSchemeProducer(value: .Success(initialValue))
-					.concat(buildProgress)
-			}
-			.startWithSignal { signal, signalDisposable in
-				disposable += signalDisposable
-				signal.observe(observer)
-			}
+					.filter { taskEvent in
+						return taskEvent.value == nil
+					}
+				
+				let versionFileSignal: Signal<TaskEvent<(ProjectLocator, String)>, CarthageError> = signal
+						.ignoreTaskData()
+						.collect()
+						.on(next: { (urls: [NSURL]) in
+							guard let dependency = dependency, rootDirectoryURL = rootDirectoryURL else {
+								return
+							}
+							if !createVersionFileForDependency(dependency, forPlatforms: options.platforms, buildProductURLs: urls, rootDirectoryURL: rootDirectoryURL) {
+								NSLog("Warning: Version file could not be created for \(dependency.project.name)")
+							}
+						})
+						.map { (urls: [NSURL]) -> TaskEvent<(ProjectLocator, String)> in
+							return .Success(ignoredValue)
+						}
+						.filter { taskEvent in
+							return false
+						}
+				
+				Signal.merge([eventSignal, versionFileSignal]).observe(observer)
```

Comment:
I'm still not sure flatMap can solve our problem here. The issue is that one stream (the `versionFileSignal`) has to wait until the end to perform some action: mainly writing out the version file. Notice the call to 'collect' midway through the definition for it.

The `eventSignal` on the other hand needs to be sending it's events upstream in real time, as the Task events contain the xcode stdout. Basically we want to fork the signal, so we can collect on it, but also let it continue.  

If we just used flatMap (or flatMapTaskEvents), we can't collect on the results, we still only get one url at a time. methods like 'scan' or other "collect but also send events" didn't work because we couldn't know when it was done, and injecting a 'takeLast' just halts the whole stream again. 

The other approach we had was to subscribe to the signal, so we could get notified of when it was done, but that either requires a collect again (which is essentially this solution here), or a local array, which is what we had done before, but it created local state which was undesired. 

We may be missing a solution to our problem, and I'm all for making it more streamlined without needing to do this merge call.  In all my searching around about 'forking' or 'splitting' signals/producers, it's either something that would need to be done higher up (at the subscribe level), or taking this approach. (e.g., (ReactiveCocoa/ReactiveCocoa#2243)